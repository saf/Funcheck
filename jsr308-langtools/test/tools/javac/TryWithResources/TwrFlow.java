/*
 * @test  /nodynamiccopyright/
 * @bug 6911256 6964740
 * @author Joseph D. Darcy
 * @summary Test exception analysis of ARM blocks
 * @compile/fail/ref=TwrFlow.out -XDrawDiagnostics TwrFlow.java
 */

import java.io.IOException;
public class TwrFlow implements AutoCloseable {
    public static void main(String... args) {
        try(TwrFlow armflow = new TwrFlow()) {
            System.out.println(armflow.toString());
        } catch (IOException ioe) { // Not reachable
            throw new AssertionError("Shouldn't reach here", ioe);
        }
        // CustomCloseException should be caught or added to throws clause

        // Also check behavior on a resource expression rather than a
        // declaration.
        TwrFlow armflowexpr = new TwrFlow();
        try(armflowexpr) {
            System.out.println(armflowexpr.toString());
        } catch (IOException ioe) { // Not reachable
            throw new AssertionError("Shouldn't reach here", ioe);
        }
        // CustomCloseException should be caught or added to throws clause
    }

    /*
     * A close method, but the class is <em>not</em> Closeable or
     * AutoCloseable.
     */
    public void close() throws CustomCloseException {
        throw new CustomCloseException();
    }
}

class CustomCloseException extends Exception {}
