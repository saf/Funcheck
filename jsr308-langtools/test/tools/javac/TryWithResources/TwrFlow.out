TwrFlow.java:14:11: compiler.err.except.never.thrown.in.try: java.io.IOException
TwrFlow.java:24:11: compiler.err.except.never.thrown.in.try: java.io.IOException
TwrFlow.java:12:46: compiler.err.unreported.exception.need.to.catch.or.throw: CustomCloseException
TwrFlow.java:22:26: compiler.err.unreported.exception.need.to.catch.or.throw: CustomCloseException
4 errors
